pkg/maps/nat/stats/cell.go

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package stats
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/cilium/cilium/pkg/maps/nat"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/metrics/metric"
+	"github.com/cilium/statedb"
+	"github.com/cilium/statedb/index"
+
+	"github.com/cilium/hive/cell"
+
+	"github.com/spf13/pflag"
+)
+
+const (
+	TableName                   = "nat-stats"
+	maxNatMapStatKStoredEntries = 2048
+	minNatMapStatKStoredEntries = 1
+	natMapStatsEntriesName      = "nat-map-stats-entries"
+)
+
+var Cell = cell.Module(
+	"nat-stats",
+	"Aggregates stats for NAT maps",
+	metrics.Metric(newMetrics),
+	cell.ProvidePrivate(newTables),
+	cell.Provide(
+		func(m Metrics) natMetrics {
+			return m
+		},
+		newStats,
+	),
+	cell.Config(Config{
+		NATMapStatInterval:       30 * time.Second,
+		NatMapStatKStoredEntries: 32,
+	}),
+	cell.Invoke(func(_ *Stats) {}),
+)
+
+type Metrics struct {
+	LocalPorts metric.Vec[metric.Gauge]
+}
+
+type natMetrics interface {
+	updateLocalPorts(family nat.IPFamily, count, maxPorts int)
+}
+
+func (m Metrics) updateLocalPorts(family nat.IPFamily, count, maxPorts int) {
+	m.LocalPorts.WithLabelValues(family.String()).Set(float64(count) / float64(maxPorts))
+}
+
+func newMetrics() Metrics {
+	return Metrics{
+		LocalPorts: metric.NewGaugeVecWithLabels(metric.GaugeOpts{
+			Namespace: metrics.Namespace,
+			Help:      "Max saturation of source ports on a egress-ip/external endpoint tuple",
+			Name:      "nat_endpoint_max_connection",
+		}, metric.Labels{
+			{Name: "family", Values: metric.NewValues(nat.IPv4.String(), nat.IPv6.String())},
+		}),
+	}
+}
+
+type Config struct {
+	NATMapStatInterval       time.Duration `mapstructure:"nat-map-stats-interval"`
+	NatMapStatKStoredEntries int           `mapstructure:"nat-map-stats-entries"`
+}
+
+func (def Config) Flags(flags *pflag.FlagSet) {
+	flags.Duration("nat-map-stats-interval", def.NATMapStatInterval, "Interval upon which nat maps are iterated for stats")
+	flags.Int("nat-map-stats-entries", def.NatMapStatKStoredEntries, "Number k top stats entries to store locally in statedb")
+}
+
+var (
+	Index = statedb.Index[NatMapStats, string]{
+		Name: "byTuple",
+		FromObject: func(s NatMapStats) index.KeySet {
+			return index.StringSlice([]string{
+				s.Type + "/" + s.Proto + " " + s.EgressIP + " " +
+					s.EndpointIP + ":" + strconv.Itoa(int(s.RemotePort))})
+		},
+		FromKey: index.String,
+		Unique:  true,
+	}
+)
+
+func newTables(db *statedb.DB) (statedb.RWTable[NatMapStats], error) {
```

Comment:
Exporting this is often useful because if some new component starts using `Table[NatMapStats]`, then it can also construct the table directly in its tests without having to use this actual cell to gain access to it. Here of course this doesn't really apply as both this and the table itself are completely private to this module, but though I'd mention it as a general guideline (which reminds me I really need to create a list of guidelines...).

Though would there be any harm in exporting `Table[NatMapStats]` (not RWTable)? Would this data be useful in any other part of our apps?

Reply:
Actually, this is intended to be exported as it will be used as a data source in further work so i'll add the `statedb.RWTable[NatMapStats].ToTable`to rexport the read only table.